### PR TITLE
[#184] Added an ordered mock call log, strict expectations and argument dispatch.

### LINKED ISSUE #184: Record an ordered call log across all mocks and assert on it
## Summary

Mock call records are kept per command, so a test can ask how many times one command was called and with what arguments, but cannot assert anything about the order in which *different* commands were called. For a script whose correctness is the order of operations - clone, then configure, then push - that is the assertion that matters most, and it is currently unexpressible.

## Details

Have every mock append a line to one shared, ordered log for the duration of the test: the command name followed by its arguments in a canonical serialisation. The serialisation needs a visible marker for an empty argument, so an empty string cannot be mistaken for an absent one when reading a failure - an invisible argument is precisely the mismatch that costs the most time to diagnose.

Then add an assertion that compares an expected sequence against the actual log and reports a unified diff on mismatch, so the failure shows what ran instead of what was expected rather than only that the two differ.

Add the degenerate cases too, since both are common and neither is expressible today:

- Assert that no mocked command was called at all.
- Assert that a particular command was never called.

Provide a way to mark a noisy command as recorded-but-excluded from the comparison. Without it, a single logging or echo call inside the code under test forces every expected sequence in the suite to carry it.

## Acceptance criteria

- [ ] A single ordered call log per test, covering every mock.
- [ ] An assertion taking an expected sequence and producing a unified diff on mismatch.
- [ ] Assertions for "no mocked command was called" and "this command was never called".
- [ ] A way to exclude a command from the comparison while still mocking it.
- [ ] Empty arguments and arguments containing whitespace round-trip unambiguously through the serialisation.
- [ ] The step runner can express an expected sequence in its own DSL rather than only through the new assertion directly.
- [ ] Positive and negative test coverage, including a case where two mocked commands interleave.

### LINKED ISSUE #185: Fail a test when a mock expectation goes unused or an unexpected call arrives
## Summary

Mocks are permissive in both directions, and both directions let a test rot into one that verifies nothing.

A mock that is declared and never called does not fail anything, so a test can assert nothing while appearing to pass. Conversely, a command invoked with arguments the test never anticipated is answered with the default status and output, so the script silently takes a path the test does not know about.

## Details

Track declared expectations per mock and verify them when the test finishes:

- An expectation that was never satisfied fails the test, with a message naming it.
- A call that matches no configured expectation fails loudly and prints the actual invocation, rather than falling back to a default response.

Guard the typo class as well: asserting against a mock name that was never registered currently passes trivially, because there are no recorded calls to contradict it. It should fail and tell the caller to register the mock first.

Make the strictness controllable so existing permissive tests have a migration path, but default new mocks to strict - a permissive default is how this gap persists.

The step runner should enforce the same rules over its declared step sequence: a declared step that is never consumed is the same bug in a different shape.

## Acceptance criteria

- [ ] An unused declared expectation fails the test with a message naming it.
- [ ] An unmatched call fails and prints the actual command line and arguments.
- [ ] Asserting against an unregistered mock fails with a message telling the caller to register it first.
- [ ] A documented escape hatch for deliberately permissive mocks.
- [ ] The step runner applies the same rules to its declared steps.
- [ ] Positive and negative test coverage for each rule.
- [ ] The README explains the strictness default and how to opt out.

### LINKED ISSUE #187: Dispatch mock behaviour on matched arguments
## Summary

A mock's configured output and status are keyed on the call ordinal - first call, second call, and so on. One mocked command therefore cannot answer two different invocations differently except by counting calls, so the test has to know the exact order in which the script happens to invoke it. Change the script's internal ordering without changing its behaviour and the test breaks. And there is no way at all to say "when called with these arguments, respond like this".

## Details

Let a mock be configured with an argument specification that selects which response applies, alongside the existing ordinal form. The ordinal form stays, unchanged, because sequencing genuinely is the right model for some tests.

The matcher vocabulary needs to be richer than exact equality to be worth the complexity:

- Match a specific argument position.
- Match by prefix, suffix or substring.
- Match by regular expression.
- Assert an argument is merely present, without a value.
- Negate any of the above.
- A position meaning "any argument matches this", and a way to pin the argument count.

First matching specification wins, in declaration order, which keeps the resolution rule explainable. Reject duplicate specifications for the same position at configuration time rather than silently picking one.

Also worth including: a forwarding mode where calls that match no specification run the real command. That makes partial mocking possible, which is the usual reason people abandon a mocking library and hand-roll a shim.

## Acceptance criteria

- [ ] Argument-specification configuration alongside the existing per-call form, with the existing form unchanged.
- [ ] Matchers for exact, prefix, suffix, substring, regex and presence, each negatable, plus positional and any-position forms.
- [ ] Argument count can be pinned.
- [ ] First matching specification wins, in declaration order, and this is documented.
- [ ] Duplicate specifications for the same position are rejected at configuration time.
- [ ] A forwarding mode for unmatched calls.
- [ ] The failure message names the argument specification that failed to match and the arguments actually received.
- [ ] Positive and negative test coverage for every matcher type and for the negation of each.

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
Closes #184
Closes #185
Closes #187

## Summary

This PR adds an ordered call log shared across every mock in a test, argument-based dispatch that selects a mock's response by matching the call's arguments, and strict expectation verification that fails a test when a mock receives a call none of its expectations cover or when a declared expectation goes unmet. Mocks are strict by default: once a response carrying a call index or an argument specification declares an expectation on a mock, a call that expectation does not cover fails the test rather than falling through to the default status and output. Setting `BATS_HELPERS_MOCK_STRICT=0` before the first `mock_command` call makes a whole suite permissive again. The step runner grows a matching `= <call>` step type and now verifies every mock it created at the end of its assert phase. `src/mock.bash` also stops being a tracked copy of an upstream project and is restyled to the repository's own conventions, so all of the mocking code lives in that one module.

## Changes

- **Call log** - every mock appends each call to one log shared by the test, so the order in which *different* commands ran is assertable. `mock_log_quote` serialises one argument in single quotes, escaping backslash, quote, tab and newline so no two distinct arguments serialise the same way. `mock_assert_calls` asserts the full ordered sequence and reports a mismatch as a unified diff; `mock_assert_no_calls`, `mock_assert_called` and `mock_assert_not_called` cover the narrower cases, the latter two failing with a clear message when the named command was never mocked, so a typo cannot pass by describing a command that was never registered. `mock_log_exclude` keeps a noisy command such as a logger visible to `mock_log_print` and the by-name assertions while leaving it out of sequence comparisons.
- **Argument specifications** - `mock_spec_add` attaches a specification to a mock; `mock_spec_arg` constrains one argument position (a one-based index, or `*` for "some argument") with a matcher - `equals`, `starts_with`, `ends_with`, `contains`, `matches` or `present`, each accepting a `not_` prefix; `mock_spec_count` pins the argument count. A numeric position may be constrained only once per specification, rejected when the specification is configured rather than silently resolved at call time. `mock_spec_set_output`, `mock_spec_set_status` and `mock_spec_set_side_effect` set the response, written literally so a value of `-n` or one holding a backslash reaches the caller unchanged. The first specification whose matchers all accept a call wins, in declaration order, and a rejected call names every specification that turned it down alongside the arguments actually received. `mock_set_forward` runs the real command for calls no specification accepts, resolving it against `PATH` with the mock directory removed so a builtin does not resolve back into the mock.
- **Strictness** - `mock_set_strict` toggles a mock's strictness, fixed at creation time from `BATS_HELPERS_MOCK_STRICT` (default `1`). A mock gains an expectation when a response is configured with a call index or when an argument specification is added; from then on a call none of them covers is rejected unless a catch-all response (one set without a call index) exempts the mock. `mock_verify` asserts every expectation of one or more mocks - or of every mock the test created, when called with none - was met, reporting both an unmet expectation and an unexpected call in the same failure.
- `src/mock.bash` - the generated mock script sources `assert.string.bash` and `mock.bash`, records its command name and resolved strictness at creation, appends every call to the shared log, and resolves its response through a matching specification before falling back to the per-call and default responses, rejecting or forwarding whatever remains unmatched. Every path baked into that generated script is quoted with `printf %q`, because the directory holding it is consumer-supplied through `BATS_HELPERS_MOCK_TMPDIR` and could otherwise inject shell code into the mock. The module is restyled to the repository's docblock, failure-message and vertical-rhythm conventions; the mock index comes from a counter file rather than a `find` over every sidecar file; a response is written with `printf` rather than `echo -e`, so a value of `-n` is no longer swallowed as a flag; and the sandbox directory is now `bats-helpers-mock`, matching `bats-helpers-backup` in `file.bash`.
- `src/steps.bash` - a new `= <call>` step adds one call, in the call-log serialisation, to the expected sequence, asserted once via `mock_assert_calls` during the assert phase. The assert phase also calls `mock_verify` for every mock the steps created, so a step sequence fails when a mocked command is called more or fewer times than the steps declare.
- `tests/mock.bats` and `tests/steps.bats` cover the above, including two interleaved commands, every matcher and its negation, and a sandbox directory whose name holds shell metacharacters. `tests/fixtures/mock_interleaved.sh` is a fixture that calls `git` and `curl` in a fixed order for the call-log tests.
- `README.md` documents the Call log, Argument specifications and Strictness sections, extends the mock setup and assertion function tables, and adds `BATS_HELPERS_MOCK_STRICT` to the environment variable reference. `CLAUDE.md` records the one rule the single module has to carry itself: the generated mock runs in its own process without `assert.base.bash`, so the functions it reaches there must never call `flunk`.

## Upgrade notes

Three changes are visible to an existing suite:

1. Mocks are strict by default. A suite whose script calls a mock more times than it configured responses for now fails instead of receiving the default response. `BATS_HELPERS_MOCK_STRICT=0` restores the previous behaviour for a whole suite, and `mock_set_strict` does it per mock.
2. A response is written literally. `mock_set_output "${mock}" 'a\nb'` now yields a literal `a\nb` rather than two lines.
3. The mock sandbox directory is `${BATS_TEST_TMPDIR}/bats-helpers-mock`. A test asserting the old path needs updating.

## Before / After

```
BEFORE                                        AFTER
┌─────────────────────────────┐               ┌──────────────────────────────┐
│ mock "git"                  │               │ mock "git" (strict, default) │
│                             │               │                              │
│ call 1 → response for #1    │               │ call → argument spec 1?      │
│ call 2 → response for #2    │               │      → argument spec 2?      │
│ call N → default response   │               │      → response for #N?      │
│          (always answered)  │               │      → catch-all default?    │
└─────────────────────────────┘               │      → none: REJECTED, and   │
                                              │        every spec is named   │
                                              └──────────────┬───────────────┘
                                                             │ every call, always
                                                             ▼
        (no shared record of calls)           ┌──────────────────────────────┐
                                              │ shared ordered call log      │
                                              │ git 'clone' '...'            │
                                              │ curl '-s' '...'              │
                                              │ git 'checkout' 'main'        │
                                              └──────────────┬───────────────┘
                                                             │
                                                             ▼
  mock_get_call_args "${mock}" 1               mock_assert_calls \
  (one mock, one call index,                     "git 'clone' '...'" \
   no ordering across commands,                  "curl '-s' '...'" \
   no record of a mismatched call)               "git 'checkout' 'main'"

                                              mock_verify
                                              (fails on an unmet expectation
                                               or an unexpected call)
```

### CODE REVIEW

File: src/mock.bash [hunk 1]
```diff
@@ -328,9 +360,55 @@ mock_command() {
...
+
+  local name_file
+  for name_file in "${dir}"/*.name; do
+    [ -e "${name_file}" ] || continue
+    cat "${name_file}"
+    echo
+  done
+
+  return 0
+}
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/mock.bash:1562]: Fixed in 1a2bcad. 'mock_names' now reads from 'mock_paths' instead of repeating the directory resolution and the '.name' scan, so the metadata layout is known in one place.

File: src/mock.match.bash [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+
+  if [ -z "${real}" ]; then
+    echo "Command '${name}' is not available to forward to" >&2
+    exit 127
+  fi
+
+  # A builtin resolves to a bare name, which 'exec' looks up on PATH, where the
+  # mock directory would otherwise still come first.
+  PATH="${forward_path}" exec "${real}" "$@"
+}
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/mock.match.bash]: Fixed in 1a2bcad. Added a 'Returns:' section recording that the function never returns to its caller, being replaced by the real command or exiting 127 when that command is not on PATH.

**Comment by AlexSkrypnyk** [comment]
@coderabbitai full review

File: tests/mock.bats [hunk 3]
```diff
@@ -169,6 +190,18 @@ load _test_helper
...
+  export BATS_HELPERS_MOCK_TMPDIR="${BATS_TEST_TMPDIR}/\$(touch pwned) mock"
+  mkdir -p "${BATS_HELPERS_MOCK_TMPDIR}"
+  mock_curl=$(mock_command "curl")
+
+  PATH="${BATS_HELPERS_MOCK_TMPDIR}":${PATH} run curl example.com
+
+  assert_success
+  assert_file_not_exists "${BATS_TEST_TMPDIR}/pwned"
+  assert_equal "curl 'example.com'" "$(mock_log_print)"
+}
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply tests/mock.bats:207]: Fixed in c6516d4. You are right that the assertion watched the wrong directory: the injected 'touch' runs in the mock process's working directory, which it inherits from the test, and that was the bats invocation directory rather than BATS_TEST_TMPDIR. The test now cd's into BATS_TEST_TMPDIR first. I verified the test genuinely discriminates by temporarily restoring the unquoted interpolation: it fails, with the diagnostic showing the path collapsed to '${BATS_TEST_TMPDIR}/ mock/...' because the command substitution had run.

File: CLAUDE.md [hunk 4]
```diff
@@ -99,15 +99,15 @@ The cost of that trade is 56 near-identical wrappers, so none of them holds logi
...
 - **Punctuate by message kind**: assertion failures routed through `format_error` take no trailing full stop; validation and runtime errors raised by a direct `flunk` call read as sentences and end with one; `Key: value` summary labels take neither.
 
 ### Documentation Style
-File headers and function docblocks follow one house style so the library reads as one project. `src/mock.bash` is exempt from the function docblock rules for the same reason it is exempt from the failure message style: its docblocks are upstream grayhemp/bats-mock text, and restyling them widens the divergence from upstream. Its file header is not upstream text and does follow the rule below.
+File headers and function docblocks follow one house style so the library reads as one project, with no exceptions.
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply CLAUDE.md:110]: Not adopting this one. The no-blank-line-after-heading shape is the existing convention throughout CLAUDE.md: 'Library Loading', 'Function Naming', 'Variable Naming', 'Assertion Functions', 'Match Modes', 'Failure Message Style', 'Mocking System' and 'Step Runner' are all written the same way, so MD022 would fire on nine headings rather than this one. Fixing only the heading this PR happens to touch would leave the file inconsistent, and the repo runs no markdownlint in CI. Worth doing as its own sweep across the file if the rule is wanted.
